lib/Dialect/FIRRTL/FIRRTLConnectionGraph.cpp

```diff
@@ -0,0 +1,222 @@
+  } else {
+    result = dest.getDefiningOp();
+  }
+  return *result;
+}
+
+FConnectLikeIterator &FConnectLikeIterator::operator++() {
+  assert(!visited && "incrementing past the end of the iterator");
+  visited = true;
+  return *this;
+}
+
+//===----------------------------------------------------------------------===//
+//
+// ResultIterator
+//
+//===----------------------------------------------------------------------===//
+
+ResultIterator::ResultIterator(FIRRTLOperation *op)
+    : op(op), resultIndexEnd(op->getNumResults()) {
+  if (resultIndex == resultIndexEnd)
+    return;
+  auto result = op->getResult(resultIndex);
+  resultIterator = result.use_begin();
+  resultIteratorEnd = result.use_end();
+  fastforward();
+}
+
+void ResultIterator::fastforward() {
+  while (resultIterator != resultIteratorEnd &&
+         dyn_cast<FConnectLike>(resultIterator.getUser()) &&
+         resultIterator.getOperand()->getOperandNumber() == 0)
+    ++resultIterator;
+  while (resultIterator == resultIteratorEnd && resultIndex != resultIndexEnd) {
+    ++resultIndex;
+    if (resultIndex == resultIndexEnd)
+      break;
+    auto result = op->getResult(resultIndex);
+    resultIterator = result.use_begin();
+    resultIteratorEnd = result.use_end();
+    while (resultIterator != resultIteratorEnd &&
+           dyn_cast<FConnectLike>(resultIterator.getUser()) &&
+           resultIterator.getOperand()->getOperandNumber() == 0)
+      ++resultIterator;
+  }
+}
+
+bool ResultIterator::operator==(const ResultIterator &other) const {
+  if (resultIndex == resultIndexEnd)
+    return other.resultIndex == other.resultIndexEnd;
+
+  return resultIterator == other.resultIterator &&
+         resultIteratorEnd == other.resultIteratorEnd &&
+         resultIndex == other.resultIndex &&
+         resultIndexEnd == other.resultIndexEnd;
+}
+
+FIRRTLOperation &ResultIterator::operator*() const {
+  assert(resultIndex != resultIndexEnd && "tried to dereference past the end");
+  return *resultIterator->getOwner();
+}
+
+ResultIterator &ResultIterator::operator++() {
+  assert(resultIndex != resultIndexEnd &&
+         "incrementing past the end of the iterator");
+  ++resultIterator;
+  fastforward();
+  return *this;
+}
+
+//===----------------------------------------------------------------------===//
+//
+// ConnectionIterator
+//
+//===----------------------------------------------------------------------===//
+
+ConnectionIterator::ConnectionIterator(FIRRTLOperation *op, bool empty) {
+  assert(op && "op should not be null");
+  TypeSwitch<FIRRTLOperation *>(op)
+      .Case<FModuleOp>([&](auto op) {
+        iterator = empty ? FModuleOpIterator() : FModuleOpIterator(op);
+      })
+      .Case<FConnectLike>([&](auto op) {
+        iterator = empty ? FConnectLikeIterator() : FConnectLikeIterator(op);
+      })
+      .Default([&](auto *op) {
+        iterator = empty ? ResultIterator() : ResultIterator(op);
+      });
```

Comment:
Constructing the default iterator to represent the end iterator is a great idea, took me a while to realize how it works.